python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -0,0 +1,394 @@
+            genes
+
+        training_options : dict,default={}
+            The parameters for the inference scheme
+        """
+        self.K = int(K)
+        self.training_options = self._fill_training_options(training_options)
+
+    def fit(self, X, progress_bar=True):
+        """
+        Fits a model given count data X.
+
+        Parameters
+        ----------
+        X : np.array-like,shape=(N,p)
+            The count data
+        Returns
+        -------
+        self
+        """
+        td = self.training_options
+        K = self.K
+
+        N, p = X.shape
+
+        model = GaussianMixture(K)
+        model.fit(X)
+        Lambda_list_l = []
+        for i in range(K):
+            Lambda_list_l.append(
+                torch.tensor(
+                    1.2 * model.means_[i].astype(np.float32) + 0.1,
+                    requires_grad=True,
+                )
+            )
+
+        myrange = trange if progress_bar else range
+
+        pct_0 = np.mean(X == 0, axis=0)
+        pct_0 = 0.98 * pct_0 + 0.01
+        pct_0_inv = np.log(pct_0 / (1 - pct_0))
+
+        alpha_list_l = [
+            torch.tensor(pct_0_inv, requires_grad=True) for i in range(K)
+        ]
+
+        sigmoid = nn.Sigmoid()
+
+        data = torch.tensor(X.astype(np.float32))
+        pi_ = 1 / K * np.ones(K)
+        myloss = PoissonNLLLoss(log_input=False, full=True, reduction="none")
+
+        for i in myrange(td["n_iterations"]):
+            # E step compute most likely categories
+            Zprobs = np.zeros((N, K))
+            for k in range(K):
+                Lambda_k = Lambda_list_l[k]
+                alpha_k = 0.8 * sigmoid(alpha_list_l[k])  # Prob of being 0
+
+                log_likelihood_poisson = -1 * myloss(Lambda_k, data)
+                log_likelihood_poisson_w = log_likelihood_poisson + torch.log(
+                    1 - alpha_k
+                )
+
+                log_likelihood_0 = torch.log(
+                    alpha_k + (1 - alpha_k) * torch.exp(-1 * Lambda_k)
+                )
+                log_likelihood_0_b = torch.broadcast_to(
+                    log_likelihood_0, data.shape
+                )
+
+                log_likelihood_point = torch.where(
+                    data != 0, log_likelihood_poisson_w, log_likelihood_0_b
+                )
+
+                log_marginal = torch.sum(log_likelihood_point, axis=1)
+
+                z_probs_k = log_marginal + np.log(pi_[k])
+                Zprobs[:, k] = z_probs_k.detach().numpy()
+
+            Ez = np.argmax(Zprobs, axis=1)
+
+            # M step, maximize parameters
+            for k in range(K):
+                pi_[k] = np.mean(Ez == k)
+                data_sub = data[Ez == k]  # Selects relevant data
+
+                trainable_variables = [Lambda_list_l[k], alpha_list_l[k]]
+
+                optimizer = torch.optim.SGD(
+                    trainable_variables, lr=td["learning_rate"], momentum=0.8
```

Comment:
do we want to pull momentum out as a parameter?

Reply:
Yeah been lazy until now.